signer/src/network/libp2p/swarm.rs

```diff
@@ -228,25 +333,49 @@ impl<'a> SignerSwarmBuilder<'a> {
     /// Build the [`SignerSwarm`], consuming the builder.
     pub fn build(self) -> Result<SignerSwarm, SignerSwarmError> {
         let keypair: Keypair = (*self.private_key).into();
-        let swarm_config = SignerSwarmConfig {
+        let behavior_config = SignerSwarmConfig {
             enable_mdns: self.enable_mdns,
+            enable_kademlia: self.enable_kademlia,
+            enable_autonat: self.enable_autonat,
             initial_bootstrap_delay: self.initial_bootstrap_delay,
             seed_addresses: self.seed_addrs,
+            num_signers: self.num_signers,
         };
-        let behavior = SignerBehavior::new(keypair.clone(), swarm_config)?;
+        let behavior = SignerBehavior::new(keypair.clone(), behavior_config)?;
+
+        // Noise (encryption) configuration.
         let noise =
             noise::Config::new(&keypair).map_err(|e| SignerSwarmError::LibP2P(Box::new(e)))?;
-        let yamux = yamux::Config::default();
-        let swarm_config = libp2p::swarm::Config::with_tokio_executor()
-            .with_idle_connection_timeout(Duration::from_secs(60));
+
+        // Yamux (muxer) configuration.
+        let mut yamux = yamux::Config::default();
+        yamux.set_max_num_streams(MAX_SUBSTREAMS_PER_CONNECTION);
+
+        // TCP transport configuration.
         let tcp_config = tcp::Config::default().nodelay(true);
 
+        // Set the dial concurrency factor to limit the number of concurrent dialing
+        // attempts. The default is 8, but we conservatively set it to 4, allowing
+        // a signer to simultaneously attempt to connect to enough peers to become
+        // well-connected while also reducing the impact of any potential connection
+        // amplification attacks.
+        let dial_concurrency_factor = NonZeroU8::new(4)
+            .ok_or_else(|| SignerSwarmError::Generic("failed to create dial concurrency factor"))?;
```

Comment:
I would probably leave this at the defaults, mainly because I don't have a good reason to change it yet, but I suppose 4 is fine. Still, we should make this a constant and put it with the rest of our constants.

Also, I think we can do `NonZeroU8::new(4).unwrap()` in `const` contexts now.